Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,592 @@
+        }
+    } else {
+        if ((error = _evalSubtractSameNumbers(result, &left, &right, comparison)) != NSCalculationNoError) {
+            return error;
+        }
+    }
+
+    NSDecimalCompact(result);
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCompact(NSDecimal* number) {
+    if ((number == NULL) || NSDecimalIsNotANumber(number)) {
+        return;
+    }
+
+    int indexer = 0;
+    // remove all the leading and trailing 0s of the mantissa.
+    while ((indexer < number->_length) && (MANTISSA(number)[indexer] == 0)) {
+        indexer++;
+    }
+
+    // shift the mantissa by indexer
+    if (indexer > 0) {
+        memmove(&MANTISSA(number)[0], &MANTISSA(number)[indexer], sizeof(unsigned short) * (LENGTH(number) - indexer));
+        LENGTH(number) -= indexer;
+    }
+
+    // remove trailing 0s
+    indexer = LENGTH(number) - 1;
+    while ((indexer >= 0) && (MANTISSA(number)[indexer] == 0) && (EXPONENT(number) < EXPONENT_MAX)) {
+        LENGTH(number)--;
+        EXPONENT(number)++;
+        indexer--;
+    }
+
+    if (LENGTH(number) == 0) {
+        SET_TO_ZERO(number);
+    }
+}
+
+/**
+ @Status Interoperable
+*/
+NSComparisonResult NSDecimalCompare(const NSDecimal* leftOperand, const NSDecimal* rightOperand) {
+    if (leftOperand == rightOperand) {
+        return NSOrderedSame;
+    }
+
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) || (rightOperand != NULL));
+
+    if (NEGATIVE_FLAG(leftOperand) != NEGATIVE_FLAG(rightOperand)) {
+        // miss match in sign.
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    int leftValue = EXPONENT(leftOperand) + LENGTH(leftOperand);
+    int rightValue = EXPONENT(rightOperand) + LENGTH(rightOperand);
+    if (leftValue > rightValue) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedAscending : NSOrderedDescending;
+    }
+    if (leftValue < rightValue) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    // same length
+    int index = std::min(LENGTH(leftOperand), LENGTH(rightOperand)) - 1;
+    while (index >= 0) {
+        if (MANTISSA(leftOperand)[index] < MANTISSA(rightOperand)[index]) {
+            return (IS_NEGATIVE(rightOperand)) ? NSOrderedDescending : NSOrderedAscending;
+        }
+
+        if (MANTISSA(leftOperand)[index] > MANTISSA(rightOperand)[index]) {
+            return IS_NEGATIVE(rightOperand) ? NSOrderedAscending : NSOrderedDescending;
+        }
+        index--;
+    }
+
+    if (LENGTH(rightOperand) > LENGTH(leftOperand)) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+    if (LENGTH(rightOperand) < LENGTH(leftOperand)) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedAscending : NSOrderedDescending;
+    }
+
+    return NSOrderedSame;
+}
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
```

Comment:
The method is being tested via NSDecimal functions that call into it, which are being called by NSDecimalNumber methods.

---
In reply to: [75560999](https://github.com/Microsoft/WinObjC/pull/896#discussion_r75560999) [](ancestors = 75560999)